src/schematools/permissions/auth.py

```diff
@@ -0,0 +1,334 @@
+"""Authorization ruleset handling.
+
+The :class:`UserScopes` class handles whether a dataset, table or field can be accessed.
+The other classes in this module ease to retrieval of permission objects.
+"""
+from __future__ import annotations
+
+from functools import lru_cache, wraps
+from typing import Callable, Iterable, Optional
+
+import methodtools
+
+from schematools.contrib.django.models import Profile
+from schematools.types import (
+    DatasetFieldSchema,
+    DatasetSchema,
+    DatasetTableSchema,
+    Permission,
+    PermissionLevel,
+    ProfileDatasetSchema,
+    ProfileSchema,
+    ProfileTableSchema,
+)
+
+HasAllScopesFunc = Callable[..., bool]  # can't describe func(*scopes: str) -> bool
+empty_dict = {}
+
+__all__ = (
+    "UserScopes",
+    "create_scopes_check",
+)
+
+
+def create_scopes_check(*user_scopes) -> HasAllScopesFunc:
+    """Create a function that tests whether the scopes are set.
+    This function only needs to be used when features such as
+    ``request.is_authorized_for(*scopes)`` from authorization_django are not available.
+    """
+    user_scopes = set(user_scopes)  # can be empty
+    return lambda *needed: user_scopes.issuperset(needed)
+
+
+def abort_on_highest(func):
+    """Decorator to abort searching for permissions when the highest value was found."""
+
+    @wraps(func)
+    def _abort_on_highest(*args, **kwargs) -> Permission:
+        try:
+            return func(*args, **kwargs)
+        except HighestPermissionFound as e:
+            return e.permission
+
+    return _abort_on_highest
+
+
+class HighestPermissionFound(Exception):
+    def __init__(self, permission: Permission):
+        self.permission = permission
+
+
+class PermissionBag:
```

Comment:
I didn't know that a "Bag" was a specific data structure. In this case, `PermissionCollection` might be better.

I choose not to extend from a list or set because `append()` would be the only supported operation. I won't want `.extend()` to work, nor `+= ..` etc.. This is really an internal tool to make the permission collection code DRY. It aborts whenever the highest permission is found, instead of having if-checks everywhere or run needlessly through all remaining profiles.